local-gpio.c

```diff
@@ -0,0 +1,329 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <arpa/inet.h>
+#include <ctype.h>
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "cdba-server.h"
+#include "local-gpio.h"
+#include "config.h"
+
+#include <gpiod.h>
+
+enum {
+	GPIO_POWER = 0,			// Power input enable
+	GPIO_FASTBOOT_KEY,		// Usually volume key
+	GPIO_POWER_KEY,			// Key to power the device
+	GPIO_USB_DISCONNECT,		// Simulate main USB connection
+	GPIO_COUNT
+};
+
+enum {
+	GPIO_ACTIVE_HIGH = 0,
+	GPIO_ACTIVE_LOW,
+};
+
+struct local_gpio {
+	char * gpiochip_desc;
+	struct gpiod_chip *chip;
+	unsigned int gpio_present[GPIO_COUNT];
+	unsigned int gpio_offset[GPIO_COUNT];
+	unsigned int gpio_polarity[GPIO_COUNT];
+#if GPIOD_MAJOR_VERSION == 2
+	struct gpiod_line_request *gpio_line[GPIO_COUNT];
+#else
+	struct gpiod_line *gpio_line[GPIO_COUNT];
+#endif
+};
+
+static int local_gpio_device_power(struct local_gpio *local_gpio, bool on);
+static void local_gpio_device_usb(struct local_gpio *local_gpio, bool on);
+
+/*
+ * fdio_gpio parameter: <gpiod chip description>;[<gpios>;...]
+ * - gpiod chip description: "gpiochip0"
+ * - gpios: type,id,polarity
+ *   - type: POWER, FASTBOOT_KEY, POWER_KEY or USB_DISCONNECT
+ *   - offset: line offset in chip
+ *   - polarity: ACTIVE_HIGH or ACTIVE_LOW
+ *
+ * Example: gpiochip0;POWER,0,ACTIVE_LOW;FASTBOOT_KEY,1,ACTIVE_HIGH;POWER_KEY,2,ACTIVE_HIGH;USB_DISCONNECT,3,ACTIVE_LOW
```

Comment:
We already have YAML. Can we please use it for the structured data?

Reply:
I'll try to use yaml, I'll do my best^^

Reply:
Rebasing

Reply:
Done, please review !